### PR TITLE
(MINOR) `IngredientOptions` allow override of hash and thumbnail generation; image library is now a default feature

### PR DESCRIPTION
**Pull request by gpeacock**
## Changes in this pull request
IngredientOptions is now a trait with three methods to override title, hash and thumbnail generation
The blake3 hash is no longer built in, if a hash is desired it can be added via ingredientOptions.
add_thumbnails feature now controls if the image crate is included in a build, by default it is enabled
updated to latest image library
added additional file type/mime mappings for ingredients

## Checklist
- [ ] This PR represents a single feature, fix, or change.
- [ ] All applicable changes have been documented.
- [ ] Any `TO DO` items (or similar) have been entered as GitHub issues and the link to that issue has been included in a comment.

### CODE REVIEW

**Comment by gpeacock** [comment]
This is required for allowing PS to not generate thumbnails via the cai-helper
Also allows for removal or replacement of the default thumbnail generation code.
Removes blake3 ingredient hash from sdk, developers can now supply their own if needed.

File: sdk/src/manifest.rs [hunk 1]
```diff
@@ -746,6 +782,7 @@ pub(crate) mod tests {
 
         assert_eq!(manifest.format(), "image/jpeg");
         assert_eq!(manifest.title(), Some("wc_embed_test.jpg"));
+        #[cfg(feature = "add_thumbnails")]
```
Comment (scouten-adobe) [inline sdk/src/manifest.rs:785]: Again, test the feature-disabled case?

File: sdk/src/utils/mod.rs [hunk 2]
```diff
@@ -17,6 +17,7 @@ pub(crate) mod hash_utils;
 #[allow(dead_code)] // for wasm build
 pub(crate) mod patch;
 #[cfg(feature = "file_io")]
```
Comment (scouten-adobe) [inline sdk/src/utils/mod.rs:19]: I think there's a `cfg(all(feature(...)` syntax that might be shorter here.

**Review by scouten-adobe — CHANGES_REQUESTED** [review]

File: sdk/src/ingredient.rs [hunk 3]
```diff
@@ -794,7 +841,9 @@ mod tests_file_io {
 
         println!("ingredient = {}", ingredient);
         assert_eq!(ingredient.title(), "libpng-test.png");
+        #[cfg(feature = "add_thumbnails")]
```
[the remark this answers is not in the document]
Comment (gpeacock) [reply sdk/src/ingredient.rs]: It took some doing, but I added no-default-features test cases

File: sdk/src/ingredient.rs [hunk 4]
```diff
@@ -828,14 +878,15 @@ mod tests_file_io {
...
-        assert_eq!(&ingredient.title, PRERELEASE_JPEG);
-        assert_eq!(&ingredient.format, "image/jpeg");
-        assert!(ingredient.thumbnail.is_some());
-        assert!(ingredient.provenance.is_some());
-        assert!(ingredient.manifest_data.is_none());
-        assert!(ingredient.validation_status.is_some());
+        assert_eq!(ingredient.title(), PRERELEASE_JPEG);
+        assert_eq!(ingredient.format(), "image/jpeg");
+        #[cfg(feature = "add_thumbnails")]
+        assert!(ingredient.thumbnail().is_some());
```
[the remark this answers is not in the document]
Comment (gpeacock) [reply sdk/src/ingredient.rs]: added test case

File: sdk/src/manifest.rs (hunk 1 above)
Comment (scouten-adobe) [reply sdk/src/manifest.rs:785]: What should `manifest.thumbnail()` return when `add_thumbnails` feature is disabled?

File: sdk/src/ingredient.rs [hunk 5]
```diff
@@ -649,17 +660,45 @@ impl std::fmt::Display for Ingredient {
...
 }
 
-#[derive(Default)]
-/// This defines optional actions when creating [`Ingredient`]s from files.
-pub struct IngredientOptions {
-    /// This allows setting the title for the ingredient. (If `None`, then the default behavior is to use the file's name.)
-    pub title: Option<&'static str>,
+/// This defines optional operations when creating [`Ingredient`] structs from files.
+#[cfg(feature = "file_io")]
+pub trait IngredientOptions {
```
Comment (scouten-adobe) [inline sdk/src/ingredient.rs:665]: Disregard my previous comment (which I've now deleted). Allowing for overriding of hash and thumbnail generation  makes sense and justifies the change from struct to trait.

File: sdk/src/ingredient.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (scouten-adobe) [reply sdk/src/ingredient.rs]: I actually had something much simpler in mind: I was thinking you could add something like ...

```rust
#[cfg(not(feature = "add_thumbnails")]
assert!(ingredient.thumbnail().is_none()); // or whatever outcome is intended
```

… within the same test.

**Review by scouten-adobe — CHANGES_REQUESTED** [review]

File: sdk/src/utils/mod.rs (hunk 2 above)
Comment (gpeacock) [reply sdk/src/utils/mod.rs:19]: done

File: sdk/src/ingredient.rs (hunk 5 above)
Comment (gpeacock) [reply sdk/src/ingredient.rs:665]: updated

File: sdk/src/manifest.rs (hunk 1 above)
Comment (gpeacock) [reply sdk/src/manifest.rs:785]: It returns the same things it does when enabled. The add_thumbnails feature only adds the ability to auto generate jpeg and png thumbnails. What manifest. thumbnail() returns depends on the test case. In some of these test cases, thumbnails will not be generated with add_thumbnails disabled, so we need to test for None in those cases only.

File: make_test_images/src/make_test_images.rs [hunk 6]
```diff
@@ -93,6 +93,26 @@ impl Default for Config {
     }
 }
 
+/// Generate a blake3 hash over the image in path using a fixed buffer
+fn blake3_hash(path: &Path) -> Result<String> {
```
Comment (mauricefisher64) [inline make_test_images/src/make_test_images.rs:97]: Just a side note.  I added blake3 support to hash_utils so it will be generally available.

File: sdk/src/ingredient.rs [hunk 7]
```diff
@@ -649,17 +660,45 @@ impl std::fmt::Display for Ingredient {
...
+        None
+    }
+
+    /// Returns an optional thumbnail image representing the asset
+    ///
+    /// The first value is the content type of the thumbnail, i.e. image/jpeg
+    /// The second value is bytes of the thumbnail image
+    /// The default is to have no thumbnail, so you must provide an override to have a thumbnail image
+    fn thumbnail(&self, _path: &Path) -> Option<(String, Vec<u8>)> {
+        #[cfg(feature = "add_thumbnails")]
```
Comment (mauricefisher64) [inline sdk/src/ingredient.rs:687]: This can now be written as:
if cfg!(feature  = "add_thumbnails") {
do something
} else {
do something else 
}

Comment (gpeacock) [reply sdk/src/ingredient.rs:687]: Ah is that new? I didn't know that was an option. thanks.

File: make_test_images/src/make_test_images.rs [hunk 8]
```diff
@@ -93,6 +93,26 @@ impl Default for Config {
...
+    use std::fs::File;
+    use std::io::Read;
+    // Hash an input incrementally.
+    let mut hasher = blake3::Hasher::new();
+    const BUFFER_LEN: usize = 1024 * 1024;
+    let mut buffer = [0u8; BUFFER_LEN];
+    let mut file = File::open(path)?;
+    loop {
+        let read_count = file.read(&mut buffer)?;
```
[the remark this answers is not in the document]
Comment (gpeacock) [reply make_test_images/src/make_test_images.rs:106]: I'll replace this with your code once everything gets merged

File: make_test_images/src/make_test_images.rs (hunk 6 above)
Comment (gpeacock) [reply make_test_images/src/make_test_images.rs:97]: I'll call that instead of what I wrote when we get things merged.

**Review by mauricefisher64 — APPROVED** [review]
👍

File: make_test_images/src/make_test_images.rs (hunk 6 above)
Comment (gpeacock) [reply make_test_images/src/make_test_images.rs:97]: i think that means we will have a Blake3 implementation in in the SDK. The question is if I should use it by default or make it a public API to to called only when desired.

File: sdk/src/ingredient.rs (hunk 7 above)
Comment (gpeacock) [reply sdk/src/ingredient.rs:687]: updated to use this syntax

File: make_test_images/src/make_test_images.rs (hunk 6 above)
Comment (gpeacock) [reply make_test_images/src/make_test_images.rs:97]: I'l integrate with the new implementation after this PR
